src/occa/internal/io/utils.cpp

```diff
@@ -442,5 +442,65 @@ namespace occa {
       fsync(fileno(fp));
       fclose(fp);
     }
+
+    std::string tmpFilenameBelow(const std::string &filename) {
+#if (OCCA_OS & (OCCA_LINUX_OS | OCCA_MACOS_OS))
+
+      /*
+        Return a temporary filename with the same basename as filename,
+        but in a temporary subdirectory.
+
+        This seems convoluted (and it is), but many variants to create
+        temporary file(name)s are unsafe, and we cannot use mkstemp
+        because it returns a file handle, not a file name, and we need
+        to able to pass the filename to e.g. compiler invocations as a string.
+
+        Unfortunate side effect: we have an extra directory that needs
+        to be cleaned. This is handled in the renameTmpFile function, so use
+        these in pairs.
+      */
+      const std::string expFilename = io::expandFilename(filename);
+      sys::mkpath(dirname(expFilename));
+      const std::string template_ = dirname(expFilename) + "XXXXXX";
+      size_t bufferSize = template_.size();
+      char buffer[bufferSize + 1];
+      ::memcpy(buffer, template_.c_str(), bufferSize);
+      // Set null terminator
+      buffer[bufferSize] = '\0';
+      char *dirname = mkdtemp(buffer);
+      OCCA_ERROR("Failed to create temporary directory for template [" << template_ << "]: " << strerror(errno),
+                 dirname != NULL);
+      return std::string(dirname) + "/" + std::string(basename(expFilename));
+#else
+#error "TODO: What to use on Windows?"
```

Comment:
> If new_filename exists, the behavior is implementation-defined. 
https://en.cppreference.com/w/cpp/io/c/rename

Perhaps choose this when available:
> existing non-directory file: new_p is first deleted, then, without allowing other processes to observe new_p as deleted, the pathname new_p is linked to the file and old_p is unlinked from the file. Write permissions are required to both the directory that contains old_p and the directory that contains new_p. 
https://en.cppreference.com/w/cpp/filesystem/rename

Reply:
Rename is already in use for the actual renaming, the TODO was for generating the file/directory name.
But that will be platform agnostic if we don't rely on `mkdtemp` anymore.

Reply:
`std::filesystem` is C++17, so unless we want to change from C++11 I'll have to stick with `std::rename` for now.